Tools/pyco/pyco.py

```diff
@@ -453,44 +458,47 @@ class HasValue(Protocol):
 
 class Builder:
     def __init__(self):
-        self.codeobjs: list[CodeObject] = []
-        self.strings: list[String | Redirect] = []
-        self.blobs: list[BlobConstant] = []
-        self.constants: list[ComplexConstant] = []
+        self.codeobjs: Mapping[CodeObject | Redirect, Tuple(int, int)] = OrderedDict()
+        self.strings: Mapping[String | Redirect, Tuple(int, int)] = OrderedDict()
+        self.blobs: Mapping[BlobConstant | Redirect, Tuple(int, int)] = OrderedDict()
+        self.constants: Mapping[ComplexConstant | Redirect, Tuple(int, int)] = OrderedDict()
         self.locked = False
-
         self.co_strings_start = -1
 
     def lock(self):
         self.locked = True
 
-    def add(self, where: list[T], thing: T) -> int:
-        # TODO: Avoid O(N**2) lookup behavior
+    def add(self, where: list[T|Redirect], thing: T|Redirect) -> int:
         # Look for a match
-        for index, it in enumerate(where):
-            if type(it.value) is type(thing.value) and it.value == thing.value:
-                return index
+        if thing in where:
+            index, _ = where[thing]
+            return index
         # Append a new one
         index = len(where)
         assert not self.locked
-        where.append(thing)
+        where[thing] = [index, index]
         return index
 
-    def add_redirect(self, where: list[T], target: int):
-        for index in range(self.co_strings_start, len(where)):
-            # Do we have one for this co?
-            it = self.strings[index]
-            if isinstance(it, Redirect) and it.target == target:
-                return index
-        return self.add(where, Redirect(target))
+    def add_redirect(self, where: list[T], thing: T, target: int):
+        thing_index, redirect_index = where[thing]
+        assert thing_index == target
+        if redirect_index >= self.co_strings_start:
+            # we already have a redirect for Thing in this co
+            # I'm not sure this can actually happen (co_names are unique, right?)
+            assert False, "Multiple redirects for a thing in the a CodeObject"
+            return redirect_index
+        index = self.add(where, Redirect(target))
+        where[thing][1] = index
```

Comment:
I changed this to `where[thing] = (where[thing], index)` and I do not get "modification during iteration" exceptions. And I don't expect such an exception unless you are inserting or deleting a key -- but we aren't, we're just mutating a value, and that should be safe. Maybe you observed this while you had some other (now fixed) bug in your code?

If you just prefer lists, the type annotations for self.codeobjs etc. should be changed, they use Tuple. (Hm, they could use lowercase tuple since it's 3.11 code. I think pyright and VS Code can type-check code with match statements in it (mypy doesn't yet).

Either way, the type annotations for add() and add_redirect() need to be updated now that `where` is a Mapping (it could just be a dict though).

Finally. To check that the indexes generated, you can change the for-loop in preload() to this:
```py
        for i, name in enumerate(code.co_names):
            index = self.builder.add_string(name)
            if index - self.co_strings_start != i:
                raise RuntimeError(f"Name index mismatch: "
                        f"{self.co_strings_start=} {i=} {name=} {index=}")
```